crates/neon-runtime/src/napi/tsfn.rs

```diff
@@ -186,10 +188,38 @@ impl<T: Send + 'static> ThreadsafeFunction<T> {
     ) {
         let Callback { callback, data } = *Box::from_raw(data as *mut Callback<T>);
 
-        // Event loop has terminated
+        // Event loop has terminated if `null`
         let env = if env.is_null() { None } else { Some(env) };
 
-        callback(env, data);
+        // Run the user supplied callback, catching panics
+        // This is unwind safe because control is never yielded back to the caller
+        let result = catch_unwind(AssertUnwindSafe(move || callback(env, data)));
+
+        // A panic occurred
+        let panic = if let Err(panic) = result {
+            panic
+        } else {
+            return;
+        };
+
+        // Try to get the panic message from common types
+        let msg = if let Some(msg) = panic.downcast_ref::<&str>() {
+            msg
+        } else if let Some(msg) = panic.downcast_ref::<String>() {
+            &msg
+        } else {
+            "Panic occurred while executing a native function scheduled by Neon"
+        };
+
+        // Try to throw an uncaught exception
+        if let Some(env) = env {
+            if super::error::fatal_exception_from_utf8(env, msg).is_ok() {
+                return;
+            }
+        }
+
+        // Failed to trigger an `uncaughtException`; fallback to print
+        eprintln!("Failed to throw an exception for panic: {}", msg);
```

Comment:
Should we print or let this disappear? Should we consider using [`log`](https://crates.io/crates/log) so users can decide for themselves what should print?